controllers/helmchart_controller.go

```diff
@@ -426,16 +435,147 @@ func (r *HelmChartReconciler) reconcileFromTarballArtifact(ctx context.Context,
+
+			if repo != nil && repo.Spec.SecretRef != nil {
+				name := types.NamespacedName{
+					Namespace: repo.GetNamespace(),
+					Name:      repo.Spec.SecretRef.Name,
+				}
+
+				var secret corev1.Secret
+				err := r.Client.Get(ctx, name, &secret)
+				if err != nil {
+					err = fmt.Errorf("auth secret error: %w", err)
+					return sourcev1.HelmChartNotReady(chart, sourcev1.AuthenticationFailedReason, err.Error()), err
+				}
+
+				opts, cleanup, err := helm.ClientOptionsFromSecret(secret)
+				if err != nil {
+					err = fmt.Errorf("auth options error: %w", err)
+					return sourcev1.HelmChartNotReady(chart, sourcev1.AuthenticationFailedReason, err.Error()), err
+				}
+				defer cleanup()
+				clientOpts = append(opts, getter.WithTimeout(repo.GetTimeout()))
+			}
+
+			// Initialize the chart repository and load the index file
+			chartRepo, err := helm.NewChartRepository(dep.Repository, r.Getters, clientOpts)
+			if err != nil {
+				switch err.(type) {
+				case *url.Error:
+					return sourcev1.HelmChartNotReady(chart, sourcev1.URLInvalidReason, err.Error()), err
+				default:
+					return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+				}
+			}
+
+			// Download index
+			err = chartRepo.DownloadIndex()
+			if err != nil {
+				return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+			}
+
+			// Lookup the chart version in the chart repository index
+			chartVer, err := chartRepo.Get(dep.Name, dep.Version)
+			if err != nil {
+				return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+			}
+
+			// Attempt to download the chart if not already existing
+			fn := path.Join(depChartsPath, fmt.Sprintf("%s-%s.tgz", chartVer.Name, chartVer.Version))
+			if _, err = os.Stat(fn); os.IsNotExist(err) {
+				res, err := chartRepo.DownloadChart(chartVer)
+				if err != nil {
+					return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+				}
+
+				// Write chart package to a file
+				f, err := os.Create(fn)
+				if err != nil {
+					err = fmt.Errorf("unable to create chart tarball file: %w", err)
+					return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+				}
+				defer f.Close()
+				_, err = f.Write(res.Bytes())
+				if err != nil {
+					err = fmt.Errorf("unable to write chart tarball file: %w", err)
+					return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+				}
+				f.Sync()
+			}
+		}
+	}
+
 	// Return early if the revision is still the same as the current chart artifact
-	newArtifact := r.Storage.NewArtifactFor(chart.Kind, chart.ObjectMeta.GetObjectMeta(), chartMetadata.Version,
-		fmt.Sprintf("%s-%s.tgz", chartMetadata.Name, chartMetadata.Version))
+	newArtifact := r.Storage.NewArtifactFor(chart.Kind, chart.ObjectMeta.GetObjectMeta(), helmChart.Metadata.Version,
```

Comment:
Given this information is static, and will not change even if we pull in new dependencies, it should be done _before_ doing anything related to dependencies.